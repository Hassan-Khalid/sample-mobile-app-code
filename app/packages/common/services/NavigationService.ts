import * as React from 'react';
import {
  CommonActions,
  DrawerActions,
  NavigationContainerRef,
  StackActions,
  useRoute
} from '@react-navigation/native';
import { AnyType } from '../declarations/types';
import SCREENS from '../../common/constants/Screens';

export const navigationRef = React.createRef() as React.RefObject<
NavigationContainerRef<AnyType>
>;

function canNavigateBack() {
  if (!navigationRef.current) {
    return false;
  }

  return navigationRef.current.canGoBack();
}

function getCurrentScreen() {
  return navigationRef.current?.getCurrentRoute()?.name || '';
}

function navigate(name: string, params?: AnyType) {
  if (!navigationRef.current) {
    return;
  }

  navigationRef.current.navigate(name, params);
}

function push(name: string, params?: AnyType) {
  if (!navigationRef.current) {
    return;
  }

  navigationRef.current.dispatch(StackActions.push(name, params));
}

function navigateBack() {
  if (!navigationRef.current) {
    return;
  }

  navigationRef.current.goBack();
}

function openDrawer(){
  if (!navigationRef.current) {
    return;
  }
  navigationRef.current.dispatch(DrawerActions.toggleDrawer());
}

function navigateTop(routeName: string, params?: AnyType) {
  const resetAction = CommonActions.reset({
    index: 0,
    routes: [{ name: routeName, params }],
  });

  if (!navigationRef.current) {
    return;
  }

  navigationRef.current.dispatch(resetAction);
}

function isMatching(routeName: string) {
  return getCurrentScreen() === routeName;
}

function getNavigationParam(param: string) {
  if (!navigationRef.current) {
    return undefined;
  }

  const routeParams = useRoute()?.params;
  return routeParams ? routeParams[param] : undefined;
}

function getTypedParam<T>(param: string) {
  if (!navigationRef.current) {
    return undefined;
  }

  const routeParams = useRoute()?.params;
  return routeParams ? (routeParams[param] as T) : undefined;
}

function bottomTabScreen () {
  const currentScreen = getCurrentScreen();
  
  switch (currentScreen) {
    case SCREENS.WORK_ORDER_LISTING:
      return true;
    default:
      return false;
  }
}

export default {
  canNavigateBack,
  getCurrentScreen,
  getNavigationParam,
  getTypedParam,
  isMatching,
  bottomTabScreen,
  openDrawer,
  navigate,
  navigateBack,
  navigateTop,
  push,
  routeName: getCurrentScreen(),
};
