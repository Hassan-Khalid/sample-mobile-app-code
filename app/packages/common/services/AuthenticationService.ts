import StorageFacade, { StorageKeys } from '../../common/facade/StorageFacade';
import NavigationService from '../../common/services/NavigationService';
import SCREENS from '../../common/constants/Screens';

interface authData {
  token: string;
  refreshToken: string
}

class AuthenticationService {
  public static handleAuthentication = (data: authData) => {
    const { token, refreshToken } = data
    StorageFacade.setKey(StorageKeys.SESSION_TOKEN, token)
    StorageFacade.setKey(StorageKeys.REFRESH_TOKEN, refreshToken)
  }
  
  public static handleLogout = () => {
    StorageFacade.deleteKey(StorageKeys.SESSION_TOKEN);
    StorageFacade.deleteKey(StorageKeys.REFRESH_TOKEN);
   NavigationService.navigate(SCREENS.LANDING);
  }
}
export default AuthenticationService;