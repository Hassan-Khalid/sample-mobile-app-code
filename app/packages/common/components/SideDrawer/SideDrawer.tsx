import React from 'react';
import { Button, SafeAreaView, View } from 'react-native';
import { AuthenticationService } from '../../services';

const SideDrawer = () => {
  return (
    <SafeAreaView>
   <View style={{flexGrow:1, marginTop:40}}>
    <Button title={'logout'} onPress={()=>AuthenticationService.handleLogout()}/>
   </View>
    </SafeAreaView>
  );
};

export default SideDrawer;