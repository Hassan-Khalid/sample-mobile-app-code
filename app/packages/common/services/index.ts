import NavigationService from './NavigationService';
import AuthenticationService from './AuthenticationService';


export {
  NavigationService,
  AuthenticationService
}