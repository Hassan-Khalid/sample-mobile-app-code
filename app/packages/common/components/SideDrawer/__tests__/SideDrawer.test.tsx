import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import SideDrawer from '../SideDrawer'; // Assuming this is the path to your component
import { AuthenticationService } from '../../../services';

jest.mock('../../../services', () => ({
  AuthenticationService: {
    handleLogout: jest.fn(),
  },
}));

describe('SideDrawer', () => {
  it('render component', () => {
    const wrapper = render(<SideDrawer />);
    expect(wrapper.toJSON).toMatchSnapshot();
  });
  
  it('calls handleLogout when the logout button is pressed', () => {
    const { getByText } = render(<SideDrawer />);
    const logoutButton = getByText('logout');
    
    fireEvent.press(logoutButton);
    
    expect(AuthenticationService.handleLogout).toHaveBeenCalled();
  });
});
