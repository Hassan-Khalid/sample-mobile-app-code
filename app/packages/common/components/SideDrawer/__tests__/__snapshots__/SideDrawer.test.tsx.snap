// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideDrawer render component 1`] = `[Function]`;
